### ISSUE TITLE
Interface name length crashes wpa_supplicant

### ISSUE THREAD

**Issue by IGNNE** [benzea/gnome-screencast#70]
Hello, the bug that some people found when using miraclecast seems to "work" with gnome-network-displays as well: 
https://github.com/albfan/miraclecast/issues/285

tl;dr If you have an interface name that is quite long (in my case wlp0s20f2), adding the "p2p-dev-" will make it bigger than the allowed length of 15, which causes wpa_supplicant to segfault. If I change my interface name to, say, wlan0, everything works.

I don't know where to report this, it feels like it is an upstream bug somewhere between the kernel, the p2p part and whoever is naming those interfaces (systemd, I guess). If you could point me in the right direction, I'll report it there (if not already done).

For the record, I'm using wpa_supplicant v2.9.3 from Debian Testing with an Intel 9560 card.

**Comment by benzea** [comment]
Well, first question would be where the 15 character limit is coming from. I think it may even be wpa_supplicant which builds the interface name in the first place.

Most likely, a wpa_supplicant bug (searching hostap is more likely to find the upstream; not sure if they have a bugtracker). Theoretically we may need to work around it on the NetworkManager level, but that seems unlikely to me right now. But I don't remember the details of what NM does at this point.

**Comment by joakim-tjernlund** [comment]
IF_NAMESIZE = 15 , it is an Linux limitation I think

**Comment by benzea** [comment]
Right, so it will be wpa_supplicant which needs to be adjusted.

**Comment by joakim-tjernlund** [comment]
It is? I figured screencast decided on the name p2p-dev- prefix?
If not, then yes wpa_supplicant needs to truncate the name to 15 chars

**Comment by benzea** [comment]
Nope, it does not. And I don't think NetworkManager does even, it just asks for a connection to be established.

**Comment by IGNNE** [comment]
Do I understand you correctly, wpa_supplicant does not rename the interface? Systemd doesn't do it either, as far as I know.

**Comment by benzea** [comment]
This is wpa_supplicant creating the interface. The relevant code is `wpas_p2p_add_group_interface` in `p2p_supplicant.c` and maybe other locations. More specifically `wpas_p2p_get_group_ifname` which does seem to have code to handle this in principle.

I think a good idea is to increase the debug level of wpa_supplicant, and then see which interface name it tries to create. You can do this by running
```
gdbus call --system --dest fi.w1.wpa_supplicant1 --object-path /fi/w1/wpa_supplicant1 --method org.freedesktop.DBus.Properties.Set fi.w1.wpa_supplicant1 '"DebugLevel"' "<'debug'>"
```
as root.

If an interface is created but just does not work, then you might also be running into https://github.com/systemd/systemd/issues/13642

**Comment by IGNNE** [comment]
Thanks for looking into this that much, I'll try it out the next time I'll have access to a miracast TV!

**Comment by IGNNE** [comment]
Looks like wpa_supplicant does just die:
```
Dec 02 13:15:08 laptop2 dbus-daemon[1004]: [system] Successfully activated service 'fi.w1.wpa_supplicant1'
Dec 02 13:15:08 laptop2 wpa_supplicant[29082]: Successfully initialized wpa_supplicant
Dec 02 13:15:08 laptop2 NetworkManager[29061]: <info>  [1575288908.3555] supplicant: wpa_supplicant running
Dec 02 13:15:08 laptop2 wpa_supplicant[29082]: nl80211: kernel reports: Attribute failed policy validation
Dec 02 13:15:08 laptop2 wpa_supplicant[29082]: Failed to create interface p2p-dev-wlp0s20f3: -22 (Invalid argument)
Dec 02 13:15:08 laptop2 wpa_supplicant[29082]: nl80211: Failed to create a P2P Device interface p2p-dev-wlp0s20f3
Dec 02 13:15:08 laptop2 wpa_supplicant[29082]: P2P: Failed to enable P2P Device interface
Dec 02 13:15:08 laptop2 wpa_supplicant[29082]: dbus: fill_dict_with_properties dbus_interface=fi.w1.wpa_supplicant1.Interface.P2PDevice dbus_property=P2PDeviceConfig getter failed

```
Then some NetworkManager stuff about how the interface becomes available, then all wifi interfaces disappear, and then the wpa_supplicant segfault:
```
Dec 02 13:15:08 laptop2 kernel: wpa_supplicant[29082]: segfault at c40 ip 000055eef2261210 sp 00007ffd71d189d8 error 4 in wpa_supplicant[55eef20ee000+1bf000]
```
To me it looks like it is the fault of wpa_supplicant.

**Comment by benzea** [comment]
Fun stuff!

So looks like we have multiple issues at hand in `wpa_supplicant` :)

The crash is one issue; I guess something goes wrong in the internal state after the device is created. If you can, try to get a backtrace and submit it.

Now, the real issue that it actually tries to create a device called `p2p-dev-wlp0s20f3 `.

Right, and I was looking at the wrong code. The relevant code is `wpas_p2p_add_p2pdev_interface` which just builds an interface name without further checks whether it may be too long. I suspect that `wpas_p2p_get_group_iface` should possibly be updated then to just always do the `ifname_ptr = os_strrchr()` thing as the prefix may be different.

**Comment by benzea** [comment]
Have you gotten around to do an upstream bug report? I suspect the only way to report it is to write to the hostapd mailing list.

**Comment by IGNNE** [comment]
Yes, I did, but I got no answer. Looks like I am not the only one.

I read in [some news (Heise, in German)](https://www.heise.de/ct/artikel/Highlights-von-Linux-5-5-Wireguard-Fundament-und-Performance-Verbesserungen-4605827.html) that kernel 5.5 will allow for longer interface names anyway, so hopefully this problem will go away by itself :-)

**Comment by benzea** [comment]
Cool, that might indeed just solve the issue!

Closing this one, I don't think there is anything actionable here at this point.

**Comment by aureooms** [comment]
Did 5.5 fix the issue? I am on 5.6.3 and still get the `p2p-scan` error.

**Comment by martinetd** [comment]
A friend ran into this recently so looked a bit -- it doesn't look like the kernel interface name length changed at all recently?

On today's master, the code is still riddled with `IFNAMSIZ` including dev name validity check here: https://github.com/torvalds/linux/blob/e99332e7b4cda6e60f5b5916cf9943a79dbef902/net/core/dev.c#L1060
And too lazy to link at `include/uapi/linux/if.h` but that still defines it to 16 ; no change whatsoever.

Now I did read through the rest of this issue and agree it's a `wpa_supplicant` bug rather than here, but this is one of the first result looking up for the error so wanted to point it out here.
FWIW a mail was sent about this to hostap's mailing list, back in december, which got no answer: http://lists.infradead.org/pipermail/hostap/2019-December/040786.html
I guess sending a patch anyway would be the next step... It's easy enough to truncate the interface name if too long (keep first few and last letters?) but there is code relying on p2p interface name to get the normal interface back so just renaming won't work; it's not something I would be comfortable sending a patch for without testing (and I have no clue what p2p interfaces are about :D)

So I'll just leave it there and send the ball back... Good luck folks :P
